### ISSUE TITLE
Test: Shutdown Kubernetes worker node

### ISSUE THREAD

**Issue by jesuslinares** [wazuh/wazuh-kubernetes#9]
Test if everything works as expected when a Kubernetes worker node is down.

**Comment by SitoRBJ** [comment]
Hello team,

Our first test is to remove a pod where we have changed Wazuh settings. 


```
ubuntu@k8s-control-server:~$ kubectl get pods --namespace wazuh
NAME                              READY     STATUS    RESTARTS   AGE
wazuh-elasticsearch-0             1/1       Running   0          6d
wazuh-kibana-78cb4bbb7-xf4s8      1/1       Running   0          6d
wazuh-logstash-646689f76f-lcf8b   1/1       Running   0          6d
wazuh-manager-master-0            1/1       Running   0          6d
wazuh-manager-worker-0-0          1/1       Running   0          6d
wazuh-manager-worker-1-0          1/1       Running   0          6d
wazuh-nginx-57c8c65486-7crh2      1/1       Running   0          6d
```

```
ubuntu@k8s-control-server:~$ kubectl delete pods wazuh-manager-worker-0-0  --namespace wazuh
pod "wazuh-manager-worker-0-0" deleted
```

```
ubuntu@k8s-control-server:~$ kubectl get pods --namespace wazuh
NAME                              READY     STATUS        RESTARTS   AGE
wazuh-elasticsearch-0             1/1       Running       0          6d
wazuh-kibana-78cb4bbb7-xf4s8      1/1       Running       0          6d
wazuh-logstash-646689f76f-lcf8b   1/1       Running       0          6d
wazuh-manager-master-0            1/1       Running       0          6d
wazuh-manager-worker-0-0          1/1       Terminating   0          6d
wazuh-manager-worker-1-0          1/1       Running       0          6d
wazuh-nginx-57c8c65486-7crh2      1/1       Running       0          6d
```

```
ubuntu@k8s-control-server:~$ kubectl get pods --namespace wazuh
NAME                              READY     STATUS              RESTARTS   AGE
wazuh-elasticsearch-0             1/1       Running             0          6d
wazuh-kibana-78cb4bbb7-xf4s8      1/1       Running             0          6d
wazuh-logstash-646689f76f-lcf8b   1/1       Running             0          6d
wazuh-manager-master-0            1/1       Running             0          6d
wazuh-manager-worker-0-0          0/1       ContainerCreating   0          14s
wazuh-manager-worker-1-0          1/1       Running             0          6d
wazuh-nginx-57c8c65486-7crh2      1/1       Running             0          6d
```

In this case, Kubernetes detects that the StatefulSet that deploys the Wazuh worker node has one less replica than it should, so Kubernetes is in charge of creating the pod again. 

Two changes were made to test what happens when you re-create the destroyed pod.

- Change in file `/var/ossec/etc/ossec.conf`

In this case, any changes made are overwritten due to the installed configuration map. 

- Change in file `/var/ossec/etc/ossec.conf`

In this case, any changes made are retained due to the use of the mounted volume. 

Kubernetes has linked the recreated pod to its old volume. 

Regards,

Alfonso Ruiz-Bravo

**Comment by SitoRBJ** [comment]
Hello team,

We are going to proceed to shut down a Kubernetes worker node.

```
ubuntu@k8s-control-server:~$ kubectl get nodes
NAME                           STATUS    ROLES     AGE       VERSION
ip-10-0-115-59.ec2.internal    Ready     <none>    20d       v1.10.3
ip-10-0-136-248.ec2.internal   Ready     <none>    20d       v1.10.3
ip-10-0-219-155.ec2.internal   Ready     <none>    20d       v1.10.3
```



```
ubuntu@k8s-control-server:~$ kubectl describe  node ip-10-0-136-248.ec2.internal
Name:               ip-10-0-136-248.ec2.internal
. . .
Non-terminated Pods:         (4 in total)
  Namespace                  Name                            CPU Requests  CPU Limits  Memory Requests  Memory Limits
  ---------                  ----                            ------------  ----------  ---------------  -------------
  kube-system                aws-node-xrgtf                  10m (0%)      0 (0%)      0 (0%)           0 (0%)
  kube-system                kube-proxy-m79fw                100m (5%)     0 (0%)      0 (0%)           0 (0%)
  wazuh                      wazuh-kibana-78cb4bbb7-xf4s8    200m (10%)    400m (20%)  512Mi (13%)      1Gi (26%)
  wazuh                      wazuh-manager-worker-0-0        500m (25%)    1 (50%)     256Mi (6%)       512Mi (13%)
Allocated resources:
  (Total limits may be over 100 percent, i.e., overcommitted.)
  CPU Requests  CPU Limits   Memory Requests  Memory Limits
  ------------  ----------   ---------------  -------------
  810m (40%)    1400m (70%)  768Mi (19%)      1536Mi (39%)
Events:         <none>

```
In this node we have the pod that contains the worker 0 of the Wazuh cluster. 


```
ubuntu@k8s-control-server:~$ kubectl describe pod wazuh-manager-worker-0-0 --namespace wazuh
7Name:           wazuh-manager-worker-0-0
Namespace:      wazuh
Node:           ip-10-0-136-248.ec2.internal/10.0.136.248
. . .
Status:         Running
IP:             10.0.138.216
Controlled By:  StatefulSet/wazuh-manager-worker-0
Containers:
  wazuh-manager:
    Container ID:   docker://d1c6eaddcf38a14ccf7b2d78df393f39f66e35421edd3a49333a30725baa9705
    Image:          wazuh/wazuh:3.7.0_6.5.0
    Image ID:       docker-pullable://wazuh/wazuh@sha256:3c8547cb000d1d39794b76988393d4e6ab7ddc3671021b1d4379f5db797a49fc
 . . .
 Mounts:
      /etc/postfix from wazuh-manager-worker (rw)
      /var/ossec/data from wazuh-manager-worker (rw)
      /var/run/secrets/kubernetes.io/serviceaccount from default-token-ttcr2 (ro)
      /wazuh-config-mount/etc/ossec.conf from config (ro)
. . .
Volumes:
  wazuh-manager-worker:
    Type:       PersistentVolumeClaim (a reference to a PersistentVolumeClaim in the same namespace)
    ClaimName:  wazuh-manager-worker-wazuh-manager-worker-0-0
    ReadOnly:   false
. . .
Events:          <none>

```
This is its persistent volume claim:

```
ubuntu@k8s-control-server:~$ kubectl describe persistentvolumeclaim wazuh-manager-worker-wazuh-manager-worker-0-0  --namespace wazuh
Name:          wazuh-manager-worker-wazuh-manager-worker-0-0
Namespace:     wazuh
StorageClass:  gp2-encrypted-retained
Status:        Bound
Volume:        pvc-ffe7bf66-f7c4-11e8-b9b8-022ada63b4ac
Labels:        app=wazuh-manager
               node-type=worker
               sts-id=0
Annotations:   pv.kubernetes.io/bind-completed=yes
               pv.kubernetes.io/bound-by-controller=yes
               volume.beta.kubernetes.io/storage-provisioner=kubernetes.io/aws-ebs
Finalizers:    [kubernetes.io/pvc-protection]
Capacity:      10Gi
Access Modes:  RWO
Events:        <none>
```

And this is its volume:

```
ubuntu@k8s-control-server:~$ kubectl describe persistentvolume pvc-ffe7bf66-f7c4-11e8-b9b8-022ada63b4ac
Name:            pvc-ffe7bf66-f7c4-11e8-b9b8-022ada63b4ac
Labels:          failure-domain.beta.kubernetes.io/region=us-east-1
                 failure-domain.beta.kubernetes.io/zone=us-east-1b
Annotations:     kubernetes.io/createdby=aws-ebs-dynamic-provisioner
                 pv.kubernetes.io/bound-by-controller=yes
                 pv.kubernetes.io/provisioned-by=kubernetes.io/aws-ebs
Finalizers:      [kubernetes.io/pv-protection]
StorageClass:    gp2-encrypted-retained
Status:          Bound
Claim:           wazuh/wazuh-manager-worker-wazuh-manager-worker-0-0
Reclaim Policy:  Retain
Access Modes:    RWO
Capacity:        10Gi
Node Affinity:   <none>
Message:
Source:
    Type:       AWSElasticBlockStore (a Persistent Disk resource in AWS)
    VolumeID:   aws://us-east-1b/vol-03193cd7aab5adaba
    FSType:     ext4
    Partition:  0
    ReadOnly:   false
Events:         <none>
```

We stopped the node.

```
ubuntu@k8s-control-server:~$ kubectl get nodes
NAME                           STATUS    ROLES     AGE       VERSION
ip-10-0-115-59.ec2.internal    Ready     <none>    20d       v1.10.3
ip-10-0-219-155.ec2.internal   Ready     <none>    20d       v1.10.3
```
Kubernetes creates a new node, (in AWS has created a new instance, ending the instance with the fallen node) and in the new node creates the worker 0 pod. 
```
ubuntu@k8s-control-server:~$ kubectl get nodes
NAME                           STATUS    ROLES     AGE       VERSION
ip-10-0-115-59.ec2.internal    Ready     <none>    20d       v1.10.3
ip-10-0-172-120.ec2.internal   Ready     <none>    3m        v1.10.3
ip-10-0-219-155.ec2.internal   Ready     <none>    20d       v1.10.3
```
```
ubuntu@k8s-control-server:~$ kubectl get pods --namespace wazuh
NAME                              READY     STATUS    RESTARTS   AGE
wazuh-elasticsearch-0             1/1       Running   0          7d
wazuh-kibana-78cb4bbb7-vsztp      1/1       Running   0          3m
wazuh-logstash-646689f76f-lcf8b   1/1       Running   0          7d
wazuh-manager-master-0            1/1       Running   0          7d
wazuh-manager-worker-0-0          1/1       Running   0          3m
wazuh-manager-worker-1-0          1/1       Running   0          6d
wazuh-nginx-57c8c65486-7crh2      1/1       Running   0          7d
```

```
ubuntu@k8s-control-server:~$ kubectl describe  node ip-10-0-172-120.ec2.internal
Name:               ip-10-0-172-120.ec2.internal
Roles:              <none>
Labels:             beta.kubernetes.io/arch=amd64
                    beta.kubernetes.io/instance-type=t2.medium
                    beta.kubernetes.io/os=linux
                    failure-domain.beta.kubernetes.io/region=us-east-1
                    failure-domain.beta.kubernetes.io/zone=us-east-1b
                    kubernetes.io/hostname=ip-10-0-172-120
Annotations:        node.alpha.kubernetes.io/ttl=0
                    volumes.kubernetes.io/controller-managed-attach-detach=true
CreationTimestamp:  Tue, 11 Dec 2018 14:33:57 +0000
Taints:             <none>
Unschedulable:      false
Conditions:
  Type             Status  LastHeartbeatTime                 LastTransitionTime                Reason                       Message
  ----             ------  -----------------                 ------------------                ------                       -------
  OutOfDisk        False   Tue, 11 Dec 2018 14:55:09 +0000   Tue, 11 Dec 2018 14:33:57 +0000   KubeletHasSufficientDisk     kubelet has sufficient disk space available
  MemoryPressure   False   Tue, 11 Dec 2018 14:55:09 +0000   Tue, 11 Dec 2018 14:33:57 +0000   KubeletHasSufficientMemory   kubelet has sufficient memory available
  DiskPressure     False   Tue, 11 Dec 2018 14:55:09 +0000   Tue, 11 Dec 2018 14:33:57 +0000   KubeletHasNoDiskPressure     kubelet has no disk pressure
  PIDPressure      False   Tue, 11 Dec 2018 14:55:09 +0000   Tue, 11 Dec 2018 14:33:57 +0000   KubeletHasSufficientPID      kubelet has sufficient PID available
  Ready            True    Tue, 11 Dec 2018 14:55:09 +0000   Tue, 11 Dec 2018 14:34:27 +0000   KubeletReady                 kubelet is posting ready status
Addresses:
  InternalIP:  10.0.172.120
  Hostname:    ip-10-0-172-120
Capacity:
 cpu:                2
 ephemeral-storage:  20959212Ki
 hugepages-2Mi:      0
 memory:             4040024Ki
 pods:               17
Allocatable:
 cpu:                2
 ephemeral-storage:  19316009748
 hugepages-2Mi:      0
 memory:             3937624Ki
 pods:               17
System Info:
 Machine ID:                 86ba47849f354636a95a361aa7f93540
 System UUID:                EC2A77A9-FBFD-346D-4A44-621F56B001AB
 Boot ID:                    c459683e-9fe1-474d-8083-5118400bd065
 Kernel Version:             4.14.62-70.117.amzn2.x86_64
 OS Image:                   Amazon Linux 2
 Operating System:           linux
 Architecture:               amd64
 Container Runtime Version:  docker://17.6.2
 Kubelet Version:            v1.10.3
 Kube-Proxy Version:         v1.10.3
ExternalID:                  i-00941634bca0b7515
ProviderID:                  aws:///us-east-1b/i-00941634bca0b7515
Non-terminated Pods:         (3 in total)
  Namespace                  Name                        CPU Requests  CPU Limits  Memory Requests  Memory Limits
  ---------                  ----                        ------------  ----------  ---------------  -------------
  kube-system                aws-node-b2srl              10m (0%)      0 (0%)      0 (0%)           0 (0%)
  kube-system                kube-proxy-67scj            100m (5%)     0 (0%)      0 (0%)           0 (0%)
  wazuh                      wazuh-manager-worker-0-0    500m (25%)    1 (50%)     256Mi (6%)       512Mi (13%)
Allocated resources:
  (Total limits may be over 100 percent, i.e., overcommitted.)
  CPU Requests  CPU Limits  Memory Requests  Memory Limits
  ------------  ----------  ---------------  -------------
  610m (30%)    1 (50%)     256Mi (6%)       512Mi (13%)
Events:
  Type    Reason                   Age   From                                   Message
  ----    ------                   ----  ----                                   -------
  Normal  Starting                 21m   kubelet, ip-10-0-172-120.ec2.internal  Starting kubelet.
  Normal  NodeHasSufficientDisk    21m   kubelet, ip-10-0-172-120.ec2.internal  Node ip-10-0-172-120.ec2.internal status is now: NodeHasSufficientDisk
  Normal  NodeHasSufficientMemory  21m   kubelet, ip-10-0-172-120.ec2.internal  Node ip-10-0-172-120.ec2.internal status is now: NodeHasSufficientMemory
  Normal  NodeHasNoDiskPressure    21m   kubelet, ip-10-0-172-120.ec2.internal  Node ip-10-0-172-120.ec2.internal status is now: NodeHasNoDiskPressure
  Normal  NodeHasSufficientPID     21m   kubelet, ip-10-0-172-120.ec2.internal  Node ip-10-0-172-120.ec2.internal status is now: NodeHasSufficientPID
  Normal  NodeAllocatableEnforced  21m   kubelet, ip-10-0-172-120.ec2.internal  Updated Node Allocatable limit across pods
  Normal  NodeReady                20m   kubelet, ip-10-0-172-120.ec2.internal  Node ip-10-0-172-120.ec2.internal status is now: NodeReady
```

The new node contains the worker pod 0 as well as the old node. The kibana pod on the other hand has been sent to the first node. 

```
ubuntu@k8s-control-server:~$ kubectl describe  node ip-10-0-115-59.ec2.internal
Name:               ip-10-0-115-59.ec2.internal
. . .
Non-terminated Pods:         (5 in total)
  Namespace                  Name                            CPU Requests  CPU Limits  Memory Requests  Memory Limits
  ---------                  ----                            ------------  ----------  ---------------  -------------
  kube-system                aws-node-s8zbn                  10m (0%)      0 (0%)      0 (0%)           0 (0%)
  kube-system                kube-proxy-xdzxj                100m (5%)     0 (0%)      0 (0%)           0 (0%)
  wazuh                      wazuh-elasticsearch-0           500m (25%)    1 (50%)     1Gi (26%)        2Gi (53%)
  wazuh                      wazuh-kibana-78cb4bbb7-vsztp    200m (10%)    400m (20%)  512Mi (13%)      1Gi (26%)
  wazuh                      wazuh-manager-worker-1-0        500m (25%)    1 (50%)     256Mi (6%)       512Mi (13%)
Allocated resources:
  (Total limits may be over 100 percent, i.e., overcommitted.)
  CPU Requests  CPU Limits    Memory Requests  Memory Limits
  ------------  ----------    ---------------  -------------
  1310m (65%)   2400m (120%)  1792Mi (46%)     3584Mi (93%)
Events:         <none>

```

```
ubuntu@k8s-control-server:~$ kubectl describe pod wazuh-manager-worker-0-0 --namespace wazuh
Name:           wazuh-manager-worker-0-0
Namespace:      wazuh
Node:           ip-10-0-172-120.ec2.internal/10.0.172.120
. . .
Status:         Running
. . .
 wazuh-manager:
    Container ID:   docker://c6d6cc26d6f7ca14e8bdc1c88cae516b3eeb6a9a9bb2033d988e26d9464ce178
    Image:          wazuh/wazuh:3.7.0_6.5.0
    Image ID:       docker-pullable://wazuh/wazuh@sha256:3c8547cb000d1d39794b76988393d4e6ab7ddc3671021b1d4379f5db797a49fc
. . .
    Mounts:
      /etc/postfix from wazuh-manager-worker (rw)
      /var/ossec/data from wazuh-manager-worker (rw)
      /var/run/secrets/kubernetes.io/serviceaccount from default-token-ttcr2 (ro)
      /wazuh-config-mount/etc/ossec.conf from config (ro)
. . .
Volumes:
  wazuh-manager-worker:
    Type:       PersistentVolumeClaim (a reference to a PersistentVolumeClaim in the same namespace)
    ClaimName:  wazuh-manager-worker-wazuh-manager-worker-0-0
    ReadOnly:   false
. . .
Events:
  Type     Reason                  Age                From                                   Message
  ----     ------                  ----               ----                                   -------
  Warning  FailedScheduling        31m (x8 over 32m)  default-scheduler                      0/2 nodes are available: 2 node(s) had no available volume zone.
  Normal   Scheduled               30m                default-scheduler                      Successfully assigned wazuh-manager-worker-0-0 to ip-10-0-172-120.ec2.internal
  Warning  FailedAttachVolume      30m                attachdetach-controller                Multi-Attach error for volume "pvc-ffe7bf66-f7c4-11e8-b9b8-022ada63b4ac" Volume is already exclusively attached to one node and can't be attached to another
  Normal   SuccessfulMountVolume   30m                kubelet, ip-10-0-172-120.ec2.internal  MountVolume.SetUp succeeded for volume "config"
  Normal   SuccessfulMountVolume   30m                kubelet, ip-10-0-172-120.ec2.internal  MountVolume.SetUp succeeded for volume "default-token-ttcr2"
  Normal   SuccessfulAttachVolume  30m                attachdetach-controller                AttachVolume.Attach succeeded for volume "pvc-ffe7bf66-f7c4-11e8-b9b8-022ada63b4ac"
  Normal   SuccessfulMountVolume   30m                kubelet, ip-10-0-172-120.ec2.internal  MountVolume.SetUp succeeded for volume "pvc-ffe7bf66-f7c4-11e8-b9b8-022ada63b4ac"
  Normal   Pulling                 30m                kubelet, ip-10-0-172-120.ec2.internal  pulling image "wazuh/wazuh:3.7.0_6.5.0"
  Normal   Pulled                  29m                kubelet, ip-10-0-172-120.ec2.internal  Successfully pulled image "wazuh/wazuh:3.7.0_6.5.0"
  Normal   Created                 29m                kubelet, ip-10-0-172-120.ec2.internal  Created container
  Normal   Started                 29m                kubelet, ip-10-0-172-120.ec2.internal  Started container
```

- Container ID before:
**d1c6eaddcf38a14ccf7b2d78df393f39f66e35421edd3a49333a30725baa9705**

- Container ID after:
**c6d6cc26d6f7ca14e8bdc1c88cae516b3eeb6a9a9bb2033d988e26d9464ce178**

The volume claim template is the same in both cases. 

```
ubuntu@k8s-control-server:~$ kubectl describe persistentvolumeclaim wazuh-manager-worker-wazuh-manager-worker-0-0  --namespace wazuh
Name:          wazuh-manager-worker-wazuh-manager-worker-0-0
Namespace:     wazuh
StorageClass:  gp2-encrypted-retained
Status:        Bound
Volume:        pvc-ffe7bf66-f7c4-11e8-b9b8-022ada63b4ac
Labels:        app=wazuh-manager
               node-type=worker
               sts-id=0
Annotations:   pv.kubernetes.io/bind-completed=yes
               pv.kubernetes.io/bound-by-controller=yes
               volume.beta.kubernetes.io/storage-provisioner=kubernetes.io/aws-ebs
Finalizers:    [kubernetes.io/pvc-protection]
Capacity:      10Gi
Access Modes:  RWO
Events:        <none>
```

```
ubuntu@k8s-control-server:~$ kubectl describe persistentvolume pvc-ffe7bf66-f7c4-11e8-b9b8-022ada63b4ac
Name:            pvc-ffe7bf66-f7c4-11e8-b9b8-022ada63b4ac
Labels:          failure-domain.beta.kubernetes.io/region=us-east-1
                 failure-domain.beta.kubernetes.io/zone=us-east-1b
Annotations:     kubernetes.io/createdby=aws-ebs-dynamic-provisioner
                 pv.kubernetes.io/bound-by-controller=yes
                 pv.kubernetes.io/provisioned-by=kubernetes.io/aws-ebs
Finalizers:      [kubernetes.io/pv-protection]
StorageClass:    gp2-encrypted-retained
Status:          Bound
Claim:           wazuh/wazuh-manager-worker-wazuh-manager-worker-0-0
Reclaim Policy:  Retain
Access Modes:    RWO
Capacity:        10Gi
Node Affinity:   <none>
Message:
Source:
    Type:       AWSElasticBlockStore (a Persistent Disk resource in AWS)
    VolumeID:   aws://us-east-1b/vol-03193cd7aab5adaba
    FSType:     ext4
    Partition:  0
    ReadOnly:   false
Events:         <none>
```
We maintain the same volume:
- Name: **pvc-ffe7bf66-f7c4-11e8-b9b8-022ada63b4ac**
- VolumeID:   **aws://us-east-1b/vol-03193cd7aab5adaba**


Regards,

Alfonso Ruiz-Bravo

**Comment by SitoRBJ** [comment]
Hello team,

Let's proceed to remove a volume. Specifically, we are going to eliminate the volume associated with Wazuh worker 0. 

```
ubuntu@k8s-control-server:~$ kubectl get persistentvolumeclaim --namespace wazuh
NAME                                            STATUS    VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS             AGE
wazuh-elasticsearch-wazuh-elasticsearch-0       Bound     pvc-b3226ad3-f7c4-11e8-b9b8-022ada63b4ac   30Gi       RWO            gp2-encrypted-retained   7d
wazuh-manager-master-wazuh-manager-master-0     Bound     pvc-fb821971-f7c4-11e8-b9b8-022ada63b4ac   10Gi       RWO            gp2-encrypted-retained   7d
wazuh-manager-worker-wazuh-manager-worker-0-0   Bound     pvc-ffe7bf66-f7c4-11e8-b9b8-022ada63b4ac   10Gi       RWO            gp2-encrypted-retained   7d
wazuh-manager-worker-wazuh-manager-worker-1-0   Bound     pvc-024466da-f7c5-11e8-b9b8-022ada63b4ac   10Gi       RWO            gp2-encrypted-retained   7d
```

```
ubuntu@k8s-control-server:~$ kubectl get persistentvolume
NAME                                       CAPACITY   ACCESS MODES   RECLAIM POLICY   STATUS    CLAIM                                                 STORAGECLASS             REASON    AGE
pvc-024466da-f7c5-11e8-b9b8-022ada63b4ac   10Gi       RWO            Retain           Bound     wazuh/wazuh-manager-worker-wazuh-manager-worker-1-0   gp2-encrypted-retained             7d
pvc-b3226ad3-f7c4-11e8-b9b8-022ada63b4ac   30Gi       RWO            Retain           Bound     wazuh/wazuh-elasticsearch-wazuh-elasticsearch-0       gp2-encrypted-retained             7d
pvc-fb821971-f7c4-11e8-b9b8-022ada63b4ac   10Gi       RWO            Retain           Bound     wazuh/wazuh-manager-master-wazuh-manager-master-0     gp2-encrypted-retained             7d
pvc-ffe7bf66-f7c4-11e8-b9b8-022ada63b4ac   10Gi       RWO            Retain           Bound     wazuh/wazuh-manager-worker-wazuh-manager-worker-0-0   gp2-encrypted-retained             7d
```

We will remove the volume:  **pvc-ffe7bf66-f7c4-11e8-b9b8-022ada63b4ac**

And this is the information that Kubernetes shows:

```
ubuntu@k8s-control-server:~$ kubectl describe pod wazuh-manager-worker-0-0 --namespace wazuh
. . .
Events:
  Type     Reason              Age   From                     Message
  ----     ------              ----  ----                     -------
  Warning  FailedAttachVolume  7m    attachdetach-controller  AttachVolume.Attach failed for volume "pvc-ffe7bf66-f7c4-11e8-b9b8-022ada63b4ac" : InvalidVolume.NotFound: The volume 'vol-03193cd7aab5adaba' does not exist.
           status code: 400, request id: 9305b784-e8d9-4269-a288-8979d8a6e8d3
  Warning  FailedAttachVolume  7m  attachdetach-controller  AttachVolume.Attach failed for volume "pvc-ffe7bf66-f7c4-11e8-b9b8-022ada63b4ac" : InvalidVolume.NotFound: The volume 'vol-03193cd7aab5adaba' does not exist.
           status code: 400, request id: 2b234f89-a1f1-44df-a465-f5d0589e5b18
  Warning  FailedAttachVolume  7m  attachdetach-controller  AttachVolume.Attach failed for volume "pvc-ffe7bf66-f7c4-11e8-b9b8-022ada63b4ac" : InvalidVolume.NotFound: The volume 'vol-03193cd7aab5adaba' does not exist.
           status code: 400, request id: be93adc1-cd53-4e89-aff2-9eb9dc9697ce
  Warning  FailedAttachVolume  7m  attachdetach-controller  AttachVolume.Attach failed for volume "pvc-ffe7bf66-f7c4-11e8-b9b8-022ada63b4ac" : InvalidVolume.NotFound: The volume 'vol-03193cd7aab5adaba' does not exist.
           status code: 400, request id: 99b06278-11c1-406f-9334-79743720b521
  Warning  FailedAttachVolume  7m  attachdetach-controller  AttachVolume.Attach failed for volume "pvc-ffe7bf66-f7c4-11e8-b9b8-022ada63b4ac" : InvalidVolume.NotFound: The volume 'vol-03193cd7aab5adaba' does not exist.
           status code: 400, request id: 6e16f96a-bad9-4ea3-8d08-1f54d31fbfd8
  Warning  FailedAttachVolume  7m  attachdetach-controller  AttachVolume.Attach failed for volume "pvc-ffe7bf66-f7c4-11e8-b9b8-022ada63b4ac" : InvalidVolume.NotFound: The volume 'vol-03193cd7aab5adaba' does not exist.
           status code: 400, request id: 94d922d5-da0f-460f-b883-cc80e9de0c1f
  Warning  FailedAttachVolume  7m  attachdetach-controller  AttachVolume.Attach failed for volume "pvc-ffe7bf66-f7c4-11e8-b9b8-022ada63b4ac" : InvalidVolume.NotFound: The volume 'vol-03193cd7aab5adaba' does not exist.
           status code: 400, request id: 0d147b1b-855b-45c6-b67a-aa33a677bbfb
  Warning  FailedAttachVolume  6m  attachdetach-controller  AttachVolume.Attach failed for volume "pvc-ffe7bf66-f7c4-11e8-b9b8-022ada63b4ac" : InvalidVolume.NotFound: The volume 'vol-03193cd7aab5adaba' does not exist.
           status code: 400, request id: 8ed2956e-8755-49a9-96cf-2ec38c946624
  Warning  FailedAttachVolume  5m  attachdetach-controller  AttachVolume.Attach failed for volume "pvc-ffe7bf66-f7c4-11e8-b9b8-022ada63b4ac" : InvalidVolume.NotFound: The volume 'vol-03193cd7aab5adaba' does not exist.
           status code: 400, request id: 6b20bed6-1af5-49c1-9215-d4c45e3de3af
  Warning  FailedAttachVolume  1m (x2 over 3m)  attachdetach-controller  (combined from similar events): AttachVolume.Attach failed for volume "pvc-ffe7bf66-f7c4-11e8-b9b8-022ada63b4ac" : InvalidVolume.NotFound: The volume 'vol-03193cd7aab5adaba' does not exist.
           status code: 400, request id: a564f681-066e-4699-9048-a439424866e0
```

If we remove the pod that linked to that volume, it won't be created again. 

```
ubuntu@k8s-control-server:~$ kubectl get pods --namespace wazuh
NAME                              READY     STATUS              RESTARTS   AGE
wazuh-elasticsearch-0             1/1       Running             0          7d
wazuh-kibana-78cb4bbb7-vsztp      1/1       Running             0          20h
wazuh-logstash-646689f76f-lcf8b   1/1       Running             0          7d
wazuh-manager-master-0            1/1       Running             0          7d
wazuh-manager-worker-0-0          0/1       ContainerCreating   0          13m
wazuh-manager-worker-1-0          1/1       Running             0          7d
wazuh-nginx-57c8c65486-7crh2      1/1       Running             0          7d
```

```
ubuntu@k8s-control-server:~$ kubectl describe pod wazuh-manager-worker-0-0 --namespace wazuh
. . .
Events:
  Type     Reason                 Age   From                                   Message
  ----     ------                 ----  ----                                   -------
  Normal   Scheduled              8m    default-scheduler                      Successfully assigned wazuh-manager-worker-0-0 to ip-10-0-172-120.ec2.internal
  Normal   SuccessfulMountVolume  8m    kubelet, ip-10-0-172-120.ec2.internal  MountVolume.SetUp succeeded for volume "config"
  Normal   SuccessfulMountVolume  8m    kubelet, ip-10-0-172-120.ec2.internal  MountVolume.SetUp succeeded for volume "default-token-ttcr2"
  Warning  FailedAttachVolume     6m    attachdetach-controller                AttachVolume.Attach failed for volume "pvc-ffe7bf66-f7c4-11e8-b9b8-022ada63b4ac" : InvalidVolume.NotFound: The volume 'vol-03193cd7aab5adaba' does not exist.
           status code: 400, request id: 66dff662-97e9-4567-aaef-06647060db37
  Warning  FailedAttachVolume  4m  attachdetach-controller  AttachVolume.Attach failed for volume "pvc-ffe7bf66-f7c4-11e8-b9b8-022ada63b4ac" : InvalidVolume.NotFound: The volume 'vol-03193cd7aab5adaba' does not exist.
           status code: 400, request id: 0e08aa47-e44a-478a-a9fc-18bfc2d13eb8
  Warning  FailedAttachVolume  2m  attachdetach-controller  AttachVolume.Attach failed for volume "pvc-ffe7bf66-f7c4-11e8-b9b8-022ada63b4ac" : InvalidVolume.NotFound: The volume 'vol-03193cd7aab5adaba' does not exist.
           status code: 400, request id: f03d88ca-43df-43ce-903a-4378925952a4
  Warning  FailedMount         2m (x3 over 6m)  kubelet, ip-10-0-172-120.ec2.internal  Unable to mount volumes for pod "wazuh-manager-worker-0-0_wazuh(80da0840-fdfc-11e8-85d1-0ae631e06c20)": timeout expired waiting for volumes to attach or mount for pod "wazuh"/"wazuh-manager-worker-0-0". list of unmounted volumes=[wazuh-manager-worker]. list of unattached volumes=[wazuh-manager-worker config default-token-ttcr2]
  Warning  FailedAttachVolume  36s              attachdetach-controller                AttachVolume.Attach failed for volume "pvc-ffe7bf66-f7c4-11e8-b9b8-022ada63b4ac" : InvalidVolume.NotFound: The volume 'vol-03193cd7aab5adaba' does not exist.
           status code: 400, request id: b6e8c6b7-6983-4eb5-9b97-f4c24488ebde
```

An option would be to remove the persistent volume claim, so when creating the pod again, it would create a new claim and create a new volume by releasing the removed volume, allowing the pod to be created correctly. 

Regards,

Alfonso Ruiz-Bravo

**Comment by SitoRBJ** [comment]
Hello team,

As far as the master node of the Kubernetes cluster is concerned. 

_Amazon Elastic Container Service for Kubernetes (Amazon EKS) is a managed service that removes the need to manage the installation, scaling, or administration of master nodes and the etcd distributed key-value store. It provides a highly available and secure K8s control plane._

_With this new service developers need not worry about deploying Kubernetes cluster or think about configuration management, ETCD clusters, and master’s high-availability._

_Further with EKS, master and ETCD will be managed by AWS and the worker nodes will be taken care by the users._

_Amazon EKS executes the Kubernetes control plane in three availability zones to ensure high availability, and automatically detects and replaces faulty master nodes._

References:

- http://blog.totalcloud.io/quick-guide-deep-dive-amazon-eks/

- https://aws.amazon.com/es/blogs/compute/run-your-kubernetes-workloads-on-amazon-ec2-spot-instances-with-amazon-eks/

- https://aws.amazon.com/es/eks/features/

**Comment by JPLachance** [comment]
Greetings,

I just want to point out that EKS doesn't provide Kubernetes audit logs yet. See the [Kubernetes auditing documentation](https://kubernetes.io/docs/tasks/debug-application-cluster/audit/#audit-policy) for more details. That means you cannot know who did what when in the cluster. :) It might be required for compliance. The AWS team is working on sending Kubernetes audit logs in Cloudwatch, but no ETA yet. Until then, if you want a Kubernetes cluster on top of AWS, [KOPS](https://github.com/kubernetes/kops) might be your best option.

About PersistentVolume, Kubernetes before 1.12 was not super aware of the topology. I did a test in EKS. I deleted a Worker pod and EKS tried to boot the pod in another AWS zone and the PersistentVolume attachment was failing. You cannot attach a volume in us-east-1a to an instance in us-east-1b. This gets fixed in [Kubernetes 1.12](https://kubernetes.io/blog/2018/10/11/topology-aware-volume-provisioning-in-kubernetes/).

Regards,

Jean-Philippe Lachance

**Comment by SitoRBJ** [comment]
Hello @JPLachance,

Thank you very much for your contribution, all this information is well received. 

In our tests with persistent volumes, we have not yet had the case that it differs in the zones, we will continue to test this situation. 

Best regards,

Alfonso Ruiz-Bravo

**Comment by jesuslinares** [comment]
Thanks so much for the information and research @SitoRBJ and @JPLachance .

>  I did a test in EKS. I deleted a Worker pod and EKS tried to boot the pod in another AWS zone and the PersistentVolume attachment was failing. You cannot attach a volume in us-east-1a to an instance in us-east-1b. This gets fixed in Kubernetes 1.12.

Great to know. We need to include it in the documentation.
